Allow to set the displayed item count and its alignment via meta

Desour opened the pull request:
Allow to set the displayed count of an item as any string and its ~~offset~~ alignment with item metadata.
Meta fields "count_meta" and "count_alignment" are used. Look into `lua_api.txt`.

Devtest has a new chatcommand, which can be used for testing.

Screenshot:
![screenshot_20190403_200206](https://user-images.githubusercontent.com/7613443/55501902-b6262580-564b-11e9-85b5-27c38e882259.png)

Closes: #8444

ClobberXD commented:
This is a best-of-both-worlds solution, and I really like it. I consider this PR to satisfy and close #8444.

ClobberXD commented:
This uses two meta keys `count` and `count_offset`, and the latter takes a complex value (i.e. it can be split into multiple individual values). Isn't it a better idea to add a separate method for this instead of using meta? e.g.

```lua
ItemStack:set_text_overlay({
    type = "stack_count", -- would take stack_count or custom, as value
    value = "50" -- only required for type == "custom"
})

-- ItemStack:get_text_overlay() would return the exact same table as above
```

DS-Minetest commented:
I've tested using colorized strings (with `minetest.colorize`) now. The color works but somehow the text is extremely far to the right. I'll try to fix this.

>Isn't it a better idea to add a separate method for this instead of using meta?

Maybe, but this would probably make this more complex. I'll better wait for core dev opinion.

DS-Minetest commented:
> I've tested using colorized strings (with `minetest.colorize`) now. The color works but somehow the text is extremely far to the right. I'll try to fix this.

Done.

DS-Minetest commented:
Documentation is added.

ClobberXD commented:
> Maybe, but this would probably make this more complex

Using meta for this feels somewhat hacky and isn't the best solution, keeping future extensibility in mind. While I agree that separate method(s) is relatively more complex, it's more straightforward and can allow extensibility without any changes.

DS-Minetest commented:
I'm scared that if I extend this too much, it will take years for this to be merged or closed and people will request more things like different text sizes, alignment to other corners than just bottom right and multiple text overlays.
Also this feature is approximately of the same sort as the other things which are done with item meta:
https://github.com/minetest/minetest/blob/00bf21e0d1a2a4a066b9d29fd5ec8184030c4435/doc/lua_api.txt#L1847-L1851

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -1849,6 +1849,10 @@ Some of the values in the key-value store are handled specially:
 * `palette_index`: If the item has a palette, this is used to get the
   current color from the palette.
+* `count`: Replace the displayed count with any string.
+* `count_offset`: Set the offset of the displayed count. This consists
```
is the default `0,0`? Worth mentioning

DS-Minetest replied:
Done.

ClobberXD commented on src/client/hud.cpp:
```diff
@@ -735,11 +735,12 @@ void drawItemStack(video::IVideoDriver *driver,
 
-	if(font != NULL && item.count >= 2)
+	const std::string &count_text = item.metadata.getString("count", 0);
+	if((font != NULL && item.count >= 2) || !count_text.empty())
```
`if ((!font &&` ...

DS-Minetest replied:
I guess you meant `if ((font &&...` (without the `!`).
I've corrected this line.

ClobberXD commented:
It might be a better idea to change the valid range of coords to [0,1], for consistency with HUD element `position` field.

DS-Minetest commented:
The comments are addressed, I think.

>change the valid range of coords to [0,1]

Negative values and values greater than 1 work, too.

LoneWolfHT commented:
I'm willing to test this if you could fix the conflicts

Desour commented:
Squashed and rebased.
@LoneWolfHT There you go.

ClobberXD commented on src/client/hud.cpp:
```diff
@@ -807,10 +807,11 @@ void drawItemStack(
+	const std::string &count_text = item.metadata.getString("count", 0);
+	if (font && (item.count >= 2 || !count_text.empty())) {
 		// Get the item count as a string
-		std::string text = itos(item.count);
-		v2u32 dim = font->getDimension(utf8_to_wide(text).c_str());
+		std::string text = count_text.empty() ? itos(item.count) : count_text;
```
If an itemstack with `count > 1` contains a non-empty `count` meta field, the stack count is replaced by the custom text in `count`. But when `count` is set to `""`, the previous non-empty value of `count` is still displayed, and nothing else happens. There's no way to revert to the stack count string at the moment.

Looking at the code, I suspect the ` || !count_text.empty()` in L811 prevents the L813 from running, which would otherwise detect that the `count` field is empty and revert to displaying the stack count.

ClobberXD commented:
Tested using the chat-commands provided in the first post. LGTM :+1:

LoneWolfHT commented:
Thanks. I tested and it seems to work. :+1:
Only thing I'd suggest is maybe a way to specify the text origin (Set it to 0.5,0.5 and the text centers on both axis. Set it to 1,1 and it's drawn like it is currently)

Desour commented:
>Only thing I'd suggest is maybe a way to specify the text origin (Set it to 0.5,0.5 and the text centers on both axis. Set it to 1,1 and it's drawn like it is currently)

Did you see that there's `count_offset`? Or do I misunderstand you?

LoneWolfHT commented:
> > Only thing I'd suggest is maybe a way to specify the text origin (Set it to 0.5,0.5 and the text centers on both axis. Set it to 1,1 and it's drawn like it is currently)
> 
> Did you see that there's `count_offset`? Or do I misunderstand you?

count_offset changes where it's drawn, count_origin would change the directions it's drawn in.
Setting `count_origin` to `0.5, 0.5` would make the text center on the point specified with `count_offset`.
Setting `count_origin` to `1, 1` would make it draw from the point specified with `count_offset` like it does right now (To the left and up)

ClobberXD commented:
I think LoneWolfHT is suggesting something equivalent to the HUD def. field `alignment`, which defines where the center of the element is.

Desour commented:
Ah, I see. I think, however, such a feature would make this PR a bit too complex.

SmallJoker commented:
The feature looks good overall but I'm concerned about scaling issues on small screens or Android. 
What would you think about fixing the text to the top left or right? It limits the proposed functionalities but ensures enough space on pretty much all screens and uniform design.
Freely adjustable offsets might result in various tools that have different modes, but all are displayed in a different spot.

ClobberXD commented:
@SmallJoker Isn't there a PR to fix HUD scaling? Wouldn't that solve scaling issues? I'm concerned about your suggestion to fix the text to the top-left (or -right) - what if I want to set a "count" for tools? My original request (#8444) was for displaying the amount of loaded ammo in a gun, while leaving the builtin tool wear for the gun's durability. Displaying it at the top will break the visual immersion, as all other items have the text at the bottom-right. If it's *really* necessary to fix the offset, I'd like it to be fixed at the bottom-right, for the sake of visual consistency.

Desour commented:
Hm, so instead of an offset it might be better to just have a finite set of positions, something like:
`right,down`(default), `right,middle`, `right,up`, `middle,down`, ...

Desour commented:
Done.
Instead of arbitrary offset positions, you can now set the alignment to one of 9 possible positions.

Desour commented:
I've made a little change:
There are now different values for the default and down/right. We might at some point want to allow the user to change the position / alignment of the displayed item count. The code no longer expects the default to be down,right. (And the order of the options is now default, left, middle, right => from up/left to down/right.)

Desour commented:
The errors and warnings in the failing build do not seem to be related to this PR's changes.

v-rob reviewed (commented):
I recall you requesting a review on IRC a while back, so... Code looks good, and seems to work pretty well.  However, is it intentional that the text can overflow the box?  It will either cover up the things below it or be obstructed itself, which can get messy:

![image](https://user-images.githubusercontent.com/31123645/116486638-ef3d1e00-a842-11eb-83c5-7498951cc24e.png)

The problem is that the text is being clipped to `clip`, but it really should be clipped to `viewrect`, which is the actual rectangle of the box.  Other than that, it works fine.

Desour commented:
You're right. Fixed. Thanks!

Desour commented:
Rebased.

Df458 reviewed (changes requested):
A couple of things I'd like fixed in here, but the overall thrust of this PR seems fine.

One thing to note is that it seems to have trouble with characters beyond ASCII, for instance "aáa" displays as "a  ", including space for the other two chars but not displaying them... not sure if we should consider that a blocker or not, I suspect this will mostly be used to add numbers where they might otherwise not be available.

One other thing: Has anyone checked if `count` is being used in item meta for other purposes by any mods/games? I highly doubt `count_meta` would be, but `count` sounds like the sort of thing that could plausibly be in use by someone. Would be good to double-check that if possible, just so we can see what this could affect.

Desour commented on src/client/hud.cpp:
```diff
@@ -1172,11 +1174,59 @@ void drawItemStack(
+				break;
+			}
+
+			rect2 = core::rect<s32>(x1, y1, x1, x2);
```
Actually, it's `x1, y1, x2, y2`. x)

Df458 reviewed (approved):
I think this is fine now. @v-rob, I saw you added the One Approval tag. Are you still fine with this now that it's been updated? If so, I think we can merge.

(As an aside, I actually meant to write "count_alignment" in my last review as a comparison rather than "count_meta" as a suggestion, but the new name is fine by me and carries less risk so I guess my mistake worked out)

Desour commented:
- Rebased.
- Added a chatcommand to devtest to print out an item.
- `rect2` was always constructed wrongly (find the difference: https://irrlicht.sourceforge.io/docu/classirr_1_1core_1_1rect.html#a8f207109cebad893de44b01cc67e9e34 ). This caused the itemcount background to never be seen. So I've fixed the constructor call and removed the background.
- Added the missing utf8 to wide conversion. (Didn't expect a stupid implicit conversion.)
- >One other thing: Has anyone checked if `count` is being used in item meta for other purposes by any mods/games?
  
  Good point.
  As meta key meaning is (contrary to item def) defined by mods, it would probably make sense to let special engine meta value keys start with `_` (ie. `_count`).
  But that would now be inconsistent with the already existing `description` and co..
  By searching in the mods in my modfolder, I found more than once that `count` was used as key for node meta.
  For now, I'll rename it to `count_meta`, as you suggested.
- (Btw. if the y-alignment is set to middle (values 8..=11), the text might appear a bit high, probably because the font dimension includes the length below the baseline. To do this better, one would need access to some of this stuff: https://en.wikipedia.org/wiki/Baseline_(typography)#/media/File:Typography_Line_Terms.svg )

wsor4035 commented:
_variable_name shouldnt be used as that is reserved for mods to use to keep them out of the engine way

wsor4035 commented:
quoted from lua_api.txt
```
        _custom_field = whatever,
        -- You can define arbitrary member variables here (see Item definition
        -- for more info) by using a '_' prefix
```

v-rob commented:
> _variable_name shouldnt be used as that is reserved for mods to use to keep them out of the engine way
>
> quoted from lua_api.txt
> 
> ```
>         _custom_field = whatever,
>         -- You can define arbitrary member variables here (see Item definition
>         -- for more info) by using a '_' prefix
> ```

That only applies to node fields, not metadata.  All the metadata I've ever seen in mods did not use underscores.  We might want to formalize something in the future.

v-rob commented:
The changes look fine to me.  I assume they're all tested, so my approval still stands.  Looks ready for merge for me.
